File: services/core/internal/inference/signvectors_test.go
package inference

import (
	"encoding/hex"
	"encoding/json"
	"os"
	"strconv"
	"testing"
)

// The Go half of the cross-language guard. The TypeScript half reads the same
// file from packages/protocol and packages/sdk.
//
// This test does not check that the layout is GOOD - the other tests do that. It
// checks that the committed vectors still describe what this code produces, so a
// layout change shows up as a failing test with an instruction rather than as a
// client whose signatures stop verifying in production.

const signVectorsPath = "../../../../packages/sdk/src/sign-vectors.json"

type signVector struct {
	Name      string `json:"name"`
	Why       string `json:"why"`
	PublicKey string `json:"publicKeyHex"`
	Provider  string `json:"provider"`
	Model     string `json:"model"`
	// A STRING because int64 nanoseconds do not survive a JSON number in
	// JavaScript. See cmd/signvectors.
	Timestamp       string              `json:"timestamp"`
	Messages        []map[string]string `json:"messages"`
	MaxTokens       int                 `json:"maxTokens"`
	Temperature     float64             `json:"temperature"`
	SigningBytesHex string              `json:"signingBytesHex"`
}

func TestTheCommittedSignVectorsMatchThisCode(t *testing.T) {
	raw, err := os.ReadFile(signVectorsPath)
	if err != nil {
		t.Fatalf("read %s: %v", signVectorsPath, err)
	}
	var vectors []signVector
	if err := json.Unmarshal(raw, &vectors); err != nil {
		t.Fatalf("parse %s: %v", signVectorsPath, err)
	}
	if len(vectors) == 0 {
		t.Fatal("the vector file is empty, so it guards nothing")
	}

	for _, v := range vectors {
		key, err := hex.DecodeString(v.PublicKey)
		if err != nil {
			t.Fatalf("%s: bad key hex: %v", v.Name, err)
		}
		msgs := make([]Message, 0, len(v.Messages))
		for _, m := range v.Messages {
			msgs = append(msgs, Message{Role: Role(m["role"]), Content: m["content"]})
		}
		ts, err := strconv.ParseInt(v.Timestamp, 10, 64)
		if err != nil {
			t.Fatalf("%s: bad timestamp %q: %v", v.Name, v.Timestamp, err)
		}
		auth := &RunAuthorization{
			PublicKey: key,
			Provider:  v.Provider,
			Model:     v.Model,
			Timestamp: ts,
		}
		got := hex.EncodeToString(auth.SigningBytes(InferenceRequest{
			Model:       v.Model,
			Messages:    msgs,
			MaxTokens:   v.MaxTokens,
			Temperature: v.Temperature,
		}))
		if got != v.SigningBytesHex {
			t.Errorf("%s (%s): the committed vector no longer matches this code.\n"+
				"  want %s\n   got %s\n"+
				"If the layout changed on purpose, regenerate and change BOTH TypeScript copies:\n"+
				"  cd services/core && go run ./cmd/signvectors > ../../packages/sdk/src/sign-vectors.json",
				v.Name, v.Why, v.SigningBytesHex, got)
		}
	}
}

// A vector file that only covered ASCII would pass while the two sides disagreed
// on every non-ASCII prompt, because that is exactly where a byte count and a
// UTF-16 count diverge. This asserts the file keeps carrying the hard cases.
func TestTheSignVectorsStillCoverTheEncodingTrap(t *testing.T) {
	raw, err := os.ReadFile(signVectorsPath)
	if err != nil {
		t.Fatalf("read %s: %v", signVectorsPath, err)
	}
	var vectors []signVector
	if err := json.Unmarshal(raw, &vectors); err != nil {
		t.Fatalf("parse %s: %v", signVectorsPath, err)
	}

	var multiByte, astral, empty bool
	for _, v := range vectors {
		fields := []string{v.Provider, v.Model}
		for _, m := range v.Messages {
			fields = append(fields, m["content"])
		}
		if len(v.Messages) == 0 {
			empty = true
		}
		for _, f := range fields {
			for _, r := range f {
				switch {
				case r > 0xFFFF:
					// Needs a surrogate PAIR in UTF-16, so the unit count and the
					// byte count differ by more than a constant factor.
					astral = true
				case r > 0x7F:
					multiByte = true
				}
			}
		}
	}
	if !multiByte {
		t.Error("no vector carries a multi-byte character, so a byte-versus-UTF-16 drift would pass")
	}
	if !astral {
		t.Error("no vector carries a character outside the BMP, which is the case a naive byte-count fix still gets wrong")
	}
	if !empty {
		t.Error("no vector has an empty transcript, which has a digest of its own")
	}
}

File: packages/sdk/src/sign-vectors.test.ts
import { describe, expect, it } from 'vitest';
import * as canonical from '@matrix-os/protocol';

import { runAuthorizationSigningBytes, type ChatMessage, type ChatRole } from './index';
import vectors from './sign-vectors.json';

/**
 * The cross-language guard: BOTH TypeScript implementations against bytes GO
 * produced.
 *
 * WHY layout-parity.test.ts DOES NOT COVER THIS. That test compares this package
 * against `@matrix-os/protocol` over randomized inputs, which catches a drift
 * between the two TypeScript copies and says nothing about Go — the side that
 * actually VERIFIES. A signature both copies compute identically and Go rejects
 * arrives in production as "invalid signature", which reads as a key problem and
 * sends somebody looking in the wrong place. On this network a run authorization
 * is what lets a browser buy inference without an API key, so that failure is
 * "nobody can pay", not "one client is odd".
 *
 * Go generates `sign-vectors.json` (`go run ./cmd/signvectors`), a Go test keeps
 * the file matching the Go code, and this test keeps TypeScript matching the file.
 * Three implementations, one source.
 *
 * THE TRAP THESE VECTORS AIM AT. Go `len()` counts BYTES; JavaScript `.length`
 * counts UTF-16 code units. They agree on ASCII and disagree on everything else,
 * so a length-prefixed layout can pass every ASCII test and still produce a
 * different digest for any prompt with an accent in it. The vectors carry Korean,
 * a character outside the BMP (a surrogate PAIR in UTF-16, where the counts differ
 * by more than a constant factor), an empty transcript and empty fields.
 */

interface SignVector {
  name: string;
  why: string;
  publicKeyHex: string;
  provider: string;
  model: string;
  /**
   * A decimal STRING, not a number.
   *
   * It is int64 nanoseconds, which passes 2^53, so a JSON number cannot carry it:
   * `1790059598000000000` parses to a different value and the digest computed from
   * it does not match Go. This test caught that on its first run against a
   * number-typed field, which is the whole reason the guard exists. Proto JSON
   * spells int64 as a string for the same reason.
   */
  timestamp: string;
  messages: { role: string; content: string }[];
  signingBytesHex: string;
}

function fromHex(hex: string): Uint8Array {
  const out = new Uint8Array(hex.length / 2);
  for (let i = 0; i < out.length; i++) out[i] = parseInt(hex.slice(i * 2, i * 2 + 2), 16);
  return out;
}

function toHex(bytes: Uint8Array): string {
  return Array.from(bytes)
    .map((b) => b.toString(16).padStart(2, '0'))
    .join('');
}

const cases = vectors as SignVector[];

/**
 * Go's roles are the internal spellings the digest is over - "user", "assistant",
 * "system". This package's public vocabulary is the PROTO enum name, which it maps
 * back with roleWireName, and that mapping is part of what a caller relies on. So
 * the vector's role is translated up into the proto name here, which tests the same
 * path `apps/web` uses rather than reaching past it.
 *
 * Worth knowing while reading this: roleWireName's default arm returns "user" for
 * anything it does not recognise. Feeding it an internal spelling therefore does
 * not fail - it silently digests an "assistant" turn as a "user" one, which is how
 * this test first appeared to have found a drift in the library. An unknown role
 * producing a valid signature over a different transcript is a sharp edge, but it
 * belongs to the published API and is not something to change from a test file.
 */
function protoRole(internal: string): ChatRole {
  switch (internal) {
    case 'system':
      return 'CHAT_ROLE_SYSTEM';
    case 'assistant':
      return 'CHAT_ROLE_ASSISTANT';
    case 'user':
      return 'CHAT_ROLE_USER';
    default:
      throw new Error(`the vectors carry a role this test does not map: ${internal}`);
  }
}

describe('run authorization bytes against the Go-generated vectors', () => {
  it('has vectors to check at all', () => {
    // A file that lost its contents would make every assertion below vacuous.
    expect(cases.length).toBeGreaterThan(0);
  });

  for (const v of cases) {
    it(`this package matches Go for ${v.name}`, async () => {
      const bytes = await runAuthorizationSigningBytes({
        fromPublicKey: fromHex(v.publicKeyHex),
        provider: v.provider,
        model: v.model,
        timestamp: BigInt(v.timestamp),
        messages: v.messages.map((m) => ({ role: protoRole(m.role), content: m.content }) as ChatMessage),
      });
      expect(toHex(bytes)).toBe(v.signingBytesHex);
    });

    it(`@matrix-os/protocol matches Go for ${v.name}`, async () => {
      const bytes = await canonical.runAuthorizationSigningBytes({
        fromPublicKey: fromHex(v.publicKeyHex),
        provider: v.provider,
        model: v.model,
        timestamp: BigInt(v.timestamp),
        messages: v.messages.map((m) => ({ role: m.role, content: m.content })) as canonical.Message[],
      });
      expect(toHex(bytes)).toBe(v.signingBytesHex);
    });
  }

  it('carries every timestamp as a string that round-trips exactly', () => {
    // The assertion that caught the original defect, kept in the form that
    // matters: a decimal string survives JSON whatever its magnitude, and a
    // number does not. If a future vector goes back to a JSON number this fails
    // rather than producing a quietly different digest.
    for (const v of cases) {
      expect(typeof v.timestamp).toBe('string');
      expect(BigInt(v.timestamp).toString()).toBe(v.timestamp);
    }
    // And at least one is past the safe-integer range, so the case that broke it
    // is still covered.
    expect(cases.some((v) => BigInt(v.timestamp) > BigInt(Number.MAX_SAFE_INTEGER))).toBe(true);
  });

  it('still covers the encoding trap, so an ASCII-only file cannot pass for a guard', () => {
    const text = cases.flatMap((v) => [v.provider, v.model, ...v.messages.map((m) => m.content)]);
    const points = text.flatMap((s) => Array.from(s).map((c) => c.codePointAt(0) ?? 0));
    expect(points.some((c) => c > 0x7f)).toBe(true);
    expect(points.some((c) => c > 0xffff)).toBe(true);
    expect(cases.some((v) => v.messages.length === 0)).toBe(true);
  });
});
